akka-stream/src/main/scala/akka/stream/StreamRefs.scala

```diff
@@ -0,0 +1,92 @@
+ * Usually obtaining a SinkRef would be done via Actor messaging, in which one system asks a remote one,
+ * to accept some data from it, and the remote one decides to accept the request to send data in a back-pressured
+ * streaming fashion -- using a sink ref.
+ *
+ * To create a [[SinkRef]] you have to materialize the `Sink` that you want to obtain a reference to by attaching it to a `StreamRefs.sinkRef()`.
+ *
+ * Stream refs can be seen as Reactive Streams over network boundaries.
+ * See also [[akka.stream.SourceRef]] which is the dual of a `SinkRef`.
+ *
+ * For additional configuration see `reference.conf` as well as [[akka.stream.StreamRefAttributes]].
+ */
+trait SinkRef[In] {
+
+  /** Scala API: Get [[Sink]] underlying to this source ref. */
+  def sink(): Sink[In, NotUsed]
+  /** Java API: Get [[javadsl.Sink]] underlying to this source ref. */
+  final def getSink(): javadsl.Sink[In, NotUsed] = sink().asJava
+}
+
+/**
+ * See full documentation on [[SourceRef]].
+ */
+object SourceRef {
+  /** Implicitly converts a SourceRef to a Source. The same can be achieved by calling `.source` on the SourceRef itself. */
+  implicit def convertRefToSource[T](ref: SourceRef[T]): Source[T, NotUsed] =
+    ref.source
+}
+
+/**
+ * A SourceRef allows sharing a "reference" with others, with the main purpose of crossing a network boundary.
+ * Usually obtaining a SourceRef would be done via Actor messaging, in which one system asks a remote one,
+ * to share some data with it, and the remote one decides to do so in a back-pressured streaming fashion -- using a stream ref.
+ *
+ * To create a [[SourceRef]] you have to materialize the `Source` that you want to obtain a reference to by attaching it to a `Sink.sourceRef`.
+ *
+ * Stream refs can be seen as Reactive Streams over network boundaries.
+ * See also [[akka.stream.SinkRef]] which is the dual of a `SourceRef`.
+ *
+ * For additional configuration see `reference.conf` as well as [[akka.stream.StreamRefAttributes]].
+ */
+trait SourceRef[T] {
+  /** Scala API: Get [[Source]] underlying to this source ref. */
+  def source: Source[T, NotUsed]
+  /** Java API: Get [[javadsl.Source]] underlying to this source ref. */
+  final def getSource: javadsl.Source[T, NotUsed] = source.asJava
+}
+
+// --- exceptions ---
+
+final case class TargetRefNotInitializedYetException()
+  extends IllegalStateException("Internal remote target actor ref not yet resolved, yet attempted to send messages to it. This should not happen due to proper flow-control, please open a ticket on the issue tracker: https://github.com/akka/akka")
+
+final case class StreamRefSubscriptionTimeoutException(msg: String)
+  extends IllegalStateException(msg)
+
+final case class RemoteStreamRefActorTerminatedException(msg: String) extends RuntimeException(msg)
+final case class InvalidSequenceNumberException(expectedSeqNr: Long, gotSeqNr: Long, msg: String)
+  extends IllegalStateException(s"$msg (expected: $expectedSeqNr, got: $gotSeqNr)")
```

Comment:
Maybe add more information about what it means? (Probably a message was dropped?)